c2rust-analyze/src/main.rs

```diff
@@ -278,6 +278,35 @@ fn update_pointer_info<'tcx>(acx: &mut AnalysisCtxt<'_, 'tcx>, mir: &Body<'tcx>)
     }
 }
 
+// Walks the type `ty` and applies a function `f` to it if it's an ADT
+// `f` gets applied recursively to `ty`'s generic types and fields (if applicable)
+fn walk_args_and_fields<'tcx, F>(tcx: TyCtxt<'tcx>, ty: Ty<'tcx>, f: &mut F)
+where
+    F: FnMut(DefId) -> bool,
+{
+    // the first type encountered in the walk is `ty`, and the
+    // relevant handling for that is below, so we can skip it here
+    for arg in ty.walk().skip(1) {
+        if let GenericArgKind::Type(ty) = arg.unpack() {
+            walk_args_and_fields(tcx, ty, f);
+        }
+    }
+
+    if let TyKind::Adt(adt_def, _) = ty.kind() {
+        if !f(adt_def.did()) {
+            return;
+        }
+        for field in adt_def.all_fields() {
+            let field_ty = tcx.type_of(field.did);
+            for arg in field_ty.walk() {
+                if let GenericArgKind::Type(ty) = arg.unpack() {
+                    walk_args_and_fields(tcx, ty, f)
+                }
+            }
+        }
+    }
```

Comment:
I'm not sure the generic walking here is correct.  From what I understand, this is what we want to do:

1. If `ty` is an ADT, call `f` on its `DefId`, which represents the generic ADT.
2. Substitute the ADTs generic args into the ADT.
3. Then recurse on each of its fields.

Reply:
> a. Substitute the ADTs generic args into the ADT.
> b. Then recurse on each of its fields.

That's overkill IMO and makes things significantly more complex.  We would need to keep a separate set of visited `Ty`s in order to avoid infinite recursion without potentially skipping some structs, whereas currently we reuse the output set (a set of `DefId`s) for this instead.

I think we should walk over generics (at least in the ADT case, if we go ahead with explicit matching on the `TyKind` instead of using `walk`), which gives us an overapproximation of reachable types but should be accurate in the vast majority of cases.

Reply:
> That's overkill IMO and makes things significantly more complex. We would need to keep a separate set of visited `Ty`s in order to avoid infinite recursion without potentially skipping some structs, whereas currently we reuse the output set (a set of `DefId`s) for this instead.

Why doesn't the `f(adt_def.did())` check prevent infinite recursion?  The current way is already recursing into the ADT's fields.

Reply:
I renamed to `walk_ffi_safe_ty` i  88472f72 now that it processes `FnPtr` too

Reply:
Btw `if !f(adt_def.did())` does prevent infinite recursion in the case that the ADT is recursive (mutually or not)`. I added documentation in 79502f74

Reply:
`walk` does seem to walk `FnPtr`s anyway, so I'd like to leave the function as-is

```rust
            ty::FnPtr(sig) => {
                stack.push(sig.skip_binder().output().into());
                stack.extend(sig.skip_binder().inputs().iter().copied().rev().map(|ty| ty.into()));
            }
```

Reply:
see https://github.com/immunant/c2rust/pull/960#discussion_r1232411197 for explanation of why walk deals with GenericArgs